### PR TITLE
Put everything in terms of FreeT and ExceptT

### PR DESCRIPTION
**Pull request by Gurkenglas**
I pulled on some loose threads and this happened.

A bunch of code was reimplementing FreeT, another bunch could be made prettier with ExceptT, such as diverging indentation. Beyond that there was much refactoring, such as replacing some lens code with lens oneliners.

The alias names/structure aren't final. Maybe there should be more newtype wrapping for better errors? Sucks that type aliases aren't properly used in GHC's error messages, though.

I haven't implemented gmask for FreeT yet. I'm not sure whether it's needed/possible.

Depends on https://github.com/alanz/ghc-mod/pull/12 and https://github.com/alanz/HaRe/pull/54 because FreeT doesn't have MonadTransControl, but if that's a hassle one could work around those dependency updates.

### CODE REVIEW

File: hie-plugin-api/Haskell/Ide/Engine/PluginsIdeMonads.hs [hunk 1]
```diff
@@ -131,163 +148,56 @@ instance ToJSON IdePlugins where
...
-  modifyMTS f = lift $ lift $ lift $ modifyMTS f
+newtype IDErring m a = IDErring { getIDErring :: ExceptT IdeError m a } deriving
+  (Functor, Applicative, Monad, MonadReader r, MonadState s, MonadIO, MonadTrans, MonadBase b, MFunctor)
+instance GM.MonadIO m => GM.MonadIO (IDErring m) where liftIO = lift . GM.liftIO
+instance GM.GmEnv m => GM.GmEnv (IDErring m) where gmeAsk = lift GM.gmeAsk; gmeLocal f x = liftWith (\run -> GM.gmeLocal f $ run x) >>= restoreT . return
+instance GM.GmLog m => GM.GmLog (IDErring m) where gmlJournal = lift . GM.gmlJournal; gmlHistory = lift GM.gmlHistory; gmlClear = lift GM.gmlClear
+instance GM.GmOut m => GM.GmOut (IDErring m) where gmoAsk = lift GM.gmoAsk
+instance GM.GmState m => GM.GmState (IDErring m) where gmsGet = lift GM.gmsGet; gmsPut = lift . GM.gmsPut; gmsState = lift . GM.gmsState
+instance (Functor f, MonadFree f m) => MonadFree f (IDErring m) where wrap x = liftWith (\run -> wrap $ fmap run x) >>= restoreT . return
+
```
Comment (alanz) [inline hie-plugin-api/Haskell/Ide/Engine/PluginsIdeMonads.hs]: Please do not let lines get longer than 115 chars (preferably 80), else the github UI does not show them properly

Comment (Gurkenglas) [reply hie-plugin-api/Haskell/Ide/Engine/PluginsIdeMonads.hs]: Huh, shows fine for me. I liked hiding the obvious boilerplate way to the right.

**Comment by Gurkenglas** [comment]
(I hope you tick the squash box on merge.)

**Comment by Gurkenglas** [comment]
`UnknownErrorCode` is a better choice there than `InternalError`, right?

**Comment by Anrock** [comment]
No response for 3 months, i guess this one is dead.

**Comment by alanz** [comment]
Agree, and there was a different refactor that achieved a similar goal that landed after.

@Gurkenglas if you want to pursue this, re-open the PR

**Pull request closed without merging**